include/zephyr/bluetooth/bluetooth.h

```diff
@@ -114,6 +114,22 @@ struct bt_le_ext_adv_cb {
 	 */
 	void (*scanned)(struct bt_le_ext_adv *adv,
 			struct bt_le_ext_adv_scanned_info *info);
+
+#if defined(CONFIG_BT_PRIVACY)
+	/**
+	 * @brief The RPA validity of the advertising set has timed out.
+	 *
+	 * This callback notifies the application that the RPA validity of
+	 * the advertising set has timed out. The user can use this callback
+	 * to synchronize the advertising payload update with the RPA rotation.
+	 *
+	 * @param adv  The advertising set object.
+	 *
+	 * @return true to rotate the current RPA, or false to use it for the
+	 *         next rotation period.
+	 */
+	bool (*rpa_timed_out)(struct bt_le_ext_adv *adv);
```

Comment:
Well, in that case ``enum`` would be a better choice in my opinion. I will leave it as it is now. Let's see what other reviewers think about this.

Do you have any specific reaction in mind for the return value (other than the two current ones)?

Reply:
`int` is more typical for the `0 for success, negative value for error`, which is also used by several other callbacks in Bluetooth (a search for `int (*` yields about 60 uses of that approach).

Reply:
This is probably because most of these callbacks return error codes. I can use the int return value if you strongly prefer having it this way. I suggest the following encoding:

```
 * @retval 0: Success: rotate the current RPA.
 * @retval Negative error value: use the same RPA for the next rotation period.
```

I will wait a bit with this change to see what others think about it.

Reply:
I think an enum would make this easier to expand on in the future and we also get named values this way.

Reply:
None of the possible return values (rotate RPA or keep using same RPA) sound like errors to me, so using errno values feels unintuitive.

Reply:
Btw, I didn't realise earlier that this was the primary purpose of this callback, i.e. to instate some kind of temporary lock on RPA rotation. There are other ways to implement such a feature than a callback. Have those been considered?

We could e.g. have `bt_rpa_lock()` and `bt_rpa_unlock()` (names can be debated) APIs to suspend RPA rotation for some period of time.

Reply:
I think we can use int as a return value with the documentation draft that I have already included in the https://github.com/zephyrproject-rtos/zephyr/pull/47260#discussion_r913695278

When we identify a need for a new reaction to this return value in the Bluetooth stack, we can simply declare an enum to be used without breaking any APIs and then extend callback documentation:

```
enum {
BT_RPA_OPTION1 = 1,
BT_RPA_OPTION2,
(...)
};

(... callback documentation ...)

 * @retval BT_RPA_OPTION1: (...)
 * @retval BT_RPA_OPTION2: (...)
 * (...)
 * @retval 0: Success: rotate the current RPA.
 * @retval Negative error value: use the same RPA for the next rotation period.
```

Using a different approach, we have also the possibility to map specific error codes to new reactions without defining the new enumerator (e.g. 0 -> Rotate RPA, ENODATA -> Do not rotate the RPA, ENOSR -> New reaction A, (...)).

This is all very theoretical since I don't see any need for other stack reactions than those already defined. Having said that, I think the int return value seems like a good tradeoff for currently defined requirements. It provides a safety net in case future extension is necessary. At the same time, it does not pollute the Bluetooth API with additional structures (like enum) when only two reactions are needed (to rotate RPA or to keep using the same value).

If you have any objections, let me know

Reply:
> Btw, I didn't realise earlier that this was the primary purpose of this callback, i.e. to instate some kind of temporary lock on RPA rotation. There are other ways to implement such a feature than a callback. Have those been considered?
> 
> We could e.g. have `bt_rpa_lock()` and `bt_rpa_unlock()` (names can be debated) APIs to suspend RPA rotation for some period of time.

I did not see your response when writing my previous comment.

Well, it's not the only purpose. The other is to rotate the advertising payload in sync with the RPA.

We can consider adding a new API called ``bt_rpa_lock`` if passing this information via callback return value is not the best choice. Please note that this will make the API a bit more complex (new function). I also think that having the ``bt_rpa_unlock`` is redundant. In my opinion, RPA locking mechanism should be temporary and valid until the next RPA timeout so that the user will not accidentally freeze its RPA permanently and compromise its privacy.

We can still stick to the boolean value as it is now since the decision to rotate or not to rotate the RPA is not an error and then design dedicated API functions if we have any need for additional reactions in the future.

Reply:
Btw, in some ways this is similar to the existing `le_force_rpa_timeout()` usage in `id.c`. There the purpose is also to guarantee that the RPA doesn't change for some period of time for specific use cases (like handing out the current connectable address over some OOB channel). The way it's implemented there is to immediately expire the RPA, so that the time counting starts from the beginning, whereas here you're extending the current period without expiring in between.

Regarding return values, `bool` feels most reasonable based on what I've seen in the discussion here.

Reply:
> Btw, in some ways this is similar to the existing `le_force_rpa_timeout()` usage in `id.c`. There the purpose is also to guarantee that the RPA doesn't change for some period of time for specific use cases (like handing out the current connectable address over some OOB channel). The way it's implemented there is to immediately expire the RPA, so that the time counting starts from the beginning, whereas here you're extending the current period without expiring in between.

Yes, I am aware of this mechanism. In some application use cases, we use the ```bt_le_oob_get_local``` API to trigger rotation manually ahead of time if there is a need for it. This API will also trigger the ```rpa_timed_out``` callback if it is registered.

@Thalley, @jilu-ot, do you agree to keep the boolean as a return value for the RPA timeout callback?

Reply:
Yes, it is fine with me.

Reply:
Based on the above `bool` sounds OK